src/scholarag/app/streaming.py

```diff
@@ -84,30 +83,47 @@ async def stream_response(
     openai_client = AsyncOpenAI(api_key=api_key)
     qas.client = openai_client
 
-    token_queue: deque[str] = deque(
-        maxlen=4
-    )  # Needs to be EXACTLY the number of tokens of the separator. See https://platform.openai.com/tokenizer
-    # Some black magic.
-    raw_string = ""
     try:
-        async for chunk in qas.astream(
+        generator = qas.astream(
             query=query,
             contexts=contexts_text,
-            prompt_template=settings.generative.prompt_template.get_secret_value(),
-        ):
-            raw_string += chunk
-            if len(token_queue) == token_queue.maxlen:
-                queued_text = "".join(list(token_queue))
-                if (
-                    f"{SOURCES_SEPARATOR}:" in queued_text
-                    or ERROR_SEPARATOR in queued_text
-                ):  # Might change if we change the separator. # This condition is subject to change based on the separator we use.
-                    continue
-                yield token_queue.popleft()
-            token_queue.append(chunk)
-    except RuntimeError as e:
-        # Since finish reason is raised, it has to be recovered as such.
-        finish_reason = e.args[0]
+            system_prompt=settings.generative.system_prompt.get_secret_value(),
+        )
+        # Get the first chunk and resulting parsing
+        _, parsed = await anext(generator)
+        # While the model didn't say if it has answer or not, keep consuming
+        while parsed.get("has_answer") is None:  # type: ignore
+            _, parsed = await anext(generator)
+        # If the LLM doesn't know answer, no need to further iterate
```

Comment:
This is debatable. We could deplete the generator and get the "raw answer" to print it in the error message. Or keep it as it is and just say that it doesn't have an answer. What do you think ?